### ISSUE TITLE
Xor and erasure coding optimization techniques

### ISSUE THREAD

**Issue by gutleib** [lizardfs/lizardfs#398]
I'm getting low performance on non-replicated datasets...
What are recommendations for optimizing xor and ec goal data for high throughput?

I'm trying raid instead of jbod on chunkservers, but it seems marginally close in our setup.
Also, do those goals have read benefit like replication?

**Comment by onlyjob** [comment]
> I'm getting low performance on non-replicated datasets...

That should be expected as _xor_ and _ec_ goals are primarily for archival purposes.
That's why originally I suggested tier implementation (similar to how it is implemented in Ceph) where _ec_ chunks are not directly writeable...

**Comment by gutleib** [comment]
I see.
But why read parallelisation is not speeding up reads from ec goaled data?
I mean, if we forget the parity data, that's good'ol RAID0 stripes...

**Comment by onlyjob** [comment]
I think you are probably comparing performance to normal replicated chunks with goal 2+ where load is balanced between chunkservers so reading a chunk from any chunkserver is sufficient. All (but one) _xor_ and _ec_ chunk parts have to be retrieved from chunkservers for successful read which is more work by definition. In both cases you can think of chunkservers as RAID-0 stripes but in first example you are reading only from one chunkserver at a time while reconstruction of _xor_ and _ec_ chunks reads from two or more chunkservers.
Anyway benefits of parallelisation are most noticeable on concurrent work loads.

**Comment by onlyjob** [comment]
> if we forget the parity data, that's good'ol RAID0 stripes...

No, _xor_ and _ec_ chunks essentially work like RAID-5 or RAID-6.

**Comment by gutleib** [comment]
@onlyjob, I understand that it works like RAID-5(6,7, etc), but in better raid algorithms those too have read benefit by reading simultaneously...

What I was meaning by calling it R-0, is that to get data 1 client has to communicate with n chunkservers (or, say, disks). Reads can happen in parallel, so you would get data more or less at the same time. That would make reading as fast as n disks (with a penalty of network latency).

What I'm asking is, how is it working now?
Is my client asking (for example let's assume xor3) 

- [request for 1/4]-[latency]-[answer]-[request for 2/4]-[latency]-[answer]-[request for 3/4]-[latency]-[answer]-[request for 4/4]-[latency]-[assemble], or
- [request for 1/4]-[request for 2/4]-[request for 3/4]-[request for 4/4]-[latency]-[assemble], or
- [request for chunk]-[(request for /4s chunkserver-to-chunkserver)-(latency)-(answer)-(answer)...-(assemble)]-[answer to client]
?

Also, does hihg-throughput raid on chunkserver improve xor/ec throughput on client?
Or low latency is more important?

**Comment by onlyjob** [comment]
I hope developers could comment on details of client read patterns.

I'd say that low latency is always more important. As for RAID configuration of HDDs on chunkserver, I believe it is worthless because it adds nothing to performance (or to reliability). You see, chunkservers with multiple HDDs randomise between 'em and you can't improve on this with RAID...

**Comment by gutleib** [comment]
Well, my tests aren't accurate now, but later I might redo them in a more controlled fashion.

But so far I'm asking for other peoples' experiences.

As to randomisations... But wouldn't that limit single operation to considerably less than one disk performance?

**Comment by onlyjob** [comment]
> As to randomisations... But wouldn't that limit single operation to considerably less than one disk performance?

Comparing to continuous sequential read from unfragmented file? Maybe just a little with much greater capacity for multiple parallel access -- a something one might be using distributed storage system for in first place. :)

**Comment by gutleib** [comment]
True, but what if one needs hundreds MB/s per single operation? What would be the correct tunables? And if one would like to get them out of ec goaled setup?
Don't need many of those concurrent, maybe 1 per each 3-5 chunkservers...
